File: src/main/java/com/bigdata/spark/java/sql/ParquetMergeSchema.java
package com.bigdata.spark.java.sql;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.SQLContext;

import java.util.List;

/**
 * Created with IntelliJ IDEA.
 * Author: CuiCan
 * Date: 2017-8-16
 * Time: 14:08
 * Description: Parquet数据源之 合并元数据
 */
public class ParquetMergeSchema {

    public static void main(String[] args) {

        SparkConf conf = new SparkConf().setAppName("ParquetMergeSchema").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);
        SQLContext sqlContext = new SQLContext(sc);

        // 首先，手动创建一个DataFrame，作为学生的基本信息数据，并将其写入目录




        // 读取parquet文件中的数据，创建一个DataFrame
        DataFrame usersDF = sqlContext.read()
                .parquet("E:\\Workspace\\intellij2017\\sparkgo\\src\\main\\resources\\parquet\\users.parquet");

        // 将df注册为临时表，然后使用SQL查询需要的数据
        usersDF.show();
        usersDF.registerTempTable("user");

        DataFrame df = sqlContext.sql("select name from user");

        // 对查询出来的df，进行transformation操作，处理数据，然后打印出来
        List<String> userNames = df.javaRDD().map(row -> "Name:" + row.getString(0)).collect();
        for (String userName : userNames) {
            System.out.println("userName = " + userName);
        }

    }

}

File: src/main/scala/com/bigdata/spark/scala/sql/ParquetMergeSchema.scala
package com.bigdata.spark.scala.sql

import org.apache.spark.sql.{SQLContext, SaveMode}
import org.apache.spark.{SparkConf, SparkContext}

/**
  * Created with IntelliJ IDEA.
  * Author: CuiCan
  * Date: 2017-8-16
  * Time: 14:32
  * Description: parquet 合并元数据
  */
object ParquetMergeSchema {

  def main(args: Array[String]): Unit = {

    val conf = new SparkConf().setAppName("ParquetMergeSchema").setMaster("local")
    val sc = new SparkContext(conf)
    val sqlContext = new SQLContext(sc)

    import sqlContext.implicits._

    // 首先，创建一个DataFrame，作为学生的基本信息数据，并将其写入parquet文件
    val studentsWithNameAge = Array(("leo", 23), ("jack", 17))
    val studentsWithNameAgeDF = sc.parallelize(studentsWithNameAge).toDF("name", "age")
    studentsWithNameAgeDF.save("./tmp/students_scala", "parquet", SaveMode.Append)

    // 其次，创建另一个DataFrame，作为学生的成绩信息，并将其写入parquet文件
    val studentsWithNameGrade = Array(("marry", "A"), ("tom", "B"))
    val studentsWithNameGradeDF = sc.parallelize(studentsWithNameGrade).toDF("name", "grade")
    studentsWithNameGradeDF.save("./tmp/students_scala", "parquet", SaveMode.Append)

    // 第一个和第二个DataFrame的元数据，肯定是不一样的，需要变成三个列的合并
    // Pattern: students: DataFrame
    val students = sqlContext.read.option("mergeSchema", "true")
      .parquet("./tmp/students_scala")

    students.printSchema()
    /**
      * root
      *  |-- name: string (nullable = true)
      *  |-- age: integer (nullable = true)
      *  |-- grade: string (nullable = true)
      */
    students.show()
    /**
      * +-----+----+-----+
      * | name| age|grade|
      * +-----+----+-----+
      * |  leo|  23| null|
      * | jack|  17| null|
      * |marry|null|    A|
      * |  tom|null|    B|
      * +-----+----+-----+
      */


    // 注册临时表，处理数据
    students.registerTempTable("students")

    //TODO


  }

}
